### PR TITLE
Choose appropriate loss function (MSE if regression) during forward as the doc suggests so that audio models could do regression

### LINKED ISSUE #27862: Audio Classification fails to do regression even though the documentation says it should under certain config
### System Info

- `transformers` version: 4.35.1
- Platform: Linux-5.4.0-166-generic-x86_64-with-glibc2.35
- Python version: 3.11.5
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.4.0
- Accelerate version: 0.24.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - debug: True
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 2
        - gpu_ids: all
        - main_process_ip: 10.18.18.1
        - main_process_port: 8080
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: True
- Using distributed or parallel set-up in script?: True

### Who can help?

Seems like @sanchit-gandhi would be of help when it comes to Whisper.

In fact, this issue could be fixed easily and I have made it work on our machine by directly modifying the source codes of `transformer` library. Though I am going to create a pull request, I think I should submit an issue here still.

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

#### Code Sample

The dataset used below is private due to license. So for one who wants to reproduce, he / she might need find a suitable dataset for audio regression.

```python
#!/home/nevikw/miniconda3/envs/ml-project/bin/python

from argparse import ArgumentParser
from random import randint
import warnings

from datasets import load_dataset, Audio, Value
from transformers import (
    AutoFeatureExtractor,
    AutoModelForAudioClassification,
    TrainingArguments,
    Trainer,
    EarlyStoppingCallback,
)
import numpy as np
from sklearn.metrics import mean_squared_error


warnings.filterwarnings("ignore")

ap = ArgumentParser()
ap.add_argument("-m", "--base-model", type=str, default="openai/whisper-large-v3")
ap.add_argument("-d", "--sample-duration", type=int, default=30)
ap.add_argument("-b", "--batch-size", type=int, default=4)
ap.add_argument("-g", "--grad-accu-step", type=int, default=8)

args = ap.parse_args()

feature_extractor = AutoFeatureExtractor.from_pretrained(args.base_model)

preprocess = lambda examples: feature_extractor(
    [i["array"][(n := randint(0, len(i["array"]) - (m := min(len(i["array"]), feature_extractor.sampling_rate*args.sample_duration)))) : n + m] for i in examples["audio"]],
    sampling_rate=feature_extractor.sampling_rate,
    do_normalize=True,
)

dataset = (
    load_dataset("nevikw39/ADReSSo")
    .cast_column("audio", Audio(sampling_rate=feature_extractor.sampling_rate))
    .cast_column("mmse", Value("float"))
)
dataset["train"], dataset["valid"] = dataset["train"].train_test_split(.25).values()

mean = np.mean(dataset["train"]["mmse"])
std = np.std(dataset["train"]["mmse"])

encoded_dataset = (
    dataset
    .map(preprocess, remove_columns=["audio"], batched=True, load_from_cache_file=False)
    .map(lambda batch: {"label": (np.array(batch["mmse"]) - mean) / std}, remove_columns=["label"], batched=True, load_from_cache_file=False)
)

model = AutoModelForAudioClassification.from_pretrained(args.base_model, num_labels=1)

training_args = TrainingArguments(
    output_dir="models/" + args.base_model[args.base_model.index('/') + 1 :] + "_ADReSSo-MMSE",
    evaluation_strategy="epoch",
    save_strategy="epoch",
    per_device_train_batch_size=args.batch_size,
    per_device_eval_batch_size=args.batch_size*2,
    gradient_accumulation_steps=args.grad_accu_step,
    num_train_epochs=100,
    warmup_ratio=.05,
    logging_steps=10,
    load_best_model_at_end=True,
    metric_for_best_model="rmse",
    greater_is_better=False,
    push_to_hub_organization="NTHU-ML-2023-team19",
    push_to_hub=False,
    hub_private_repo=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=encoded_dataset["train"],
    eval_dataset=encoded_dataset["valid"],
    tokenizer=feature_extractor,
    compute_metrics=lambda eval_pred: {
        "rmse": mean_squared_error(eval_pred.label_ids, eval_pred.predictions, squared=False) * std,
    },
    callbacks=[EarlyStoppingCallback(10)],
)

trainer.train()

print(trainer.evaluate(encoded_dataset["test"]))

trainer.save_model("models/" + args.base_model[args.base_model.index('/') + 1 :] + "_ADReSSo-MMSE")
```

#### Error Message

```
Traceback (most recent call last):
  File "/home/nevikw/ML_Project/./acoustic_ft_mmse.py", line 106, in <module>
    trainer.train()
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/transformers/trainer.py", line 1555, in train
    return inner_training_loop(
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/transformers/trainer.py", line 1860, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/transformers/trainer.py", line 2725, in training_step
    loss = self.compute_loss(model, inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/transformers/trainer.py", line 2748, in compute_loss
    outputs = model(**inputs)
              ^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/parallel/data_parallel.py", line 185, in forward
    outputs = self.parallel_apply(replicas, inputs, module_kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/parallel/data_parallel.py", line 200, in parallel_apply
    return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/parallel/parallel_apply.py", line 110, in parallel_apply
    output.reraise()
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/_utils.py", line 694, in reraise
    raise exception
RuntimeError: Caught RuntimeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in _worker
    output = module(*input, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/transformers/models/whisper/modeling_whisper.py", line 2419, in forward
    loss = loss_fct(logits.view(-1, self.config.num_labels), labels.view(-1))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/modules/loss.py", line 1179, in forward
    return F.cross_entropy(input, target, weight=self.weight,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nevikw/miniconda3/envs/ml-project/lib/python3.11/site-packages/torch/nn/functional.py", line 3053, in cross_entropy
    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: "nll_loss_forward_reduce_cuda_kernel_2d_index" not implemented for 'Float'
```

#### Proposed Solution

I found that the issue could be resolved by assigning appropriate loss function to `loss_fct` in `forward()` method of `WhisperForAudioClassification` class. The pull request will be created latter.

### Expected behavior

We should be able to perform the regression task and the mean square error loss should be computed during forward process if `config.num_labels=1` as the documentation suggests.

### PR DESCRIPTION
**Pull request by nevikw39**
# What does this PR do?

The general documentation of `transformers` says that a classifier would specialize in regression if `num_labels` in config is set to 1. The doc-string comments of `WhisperForAudioClassification` class also suggest that it would do so. Nevertheless, the actual implementation of `forward()` method always computes cross entropy loss in regardless of the config.

The issue is found on other audio classification models as well, such as wav2vec2.0, HuBERT, WavLM, etc.

So this PR assigns appropriate loss function to `loss_fct` in `forward()` method of audio classification models class (mean square loss for regression) as the documentation suggests.

Fixes #27862 

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
  -  No, this PR fixed a small bug.
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
  -  Yes of course!
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link  to it if that's the case.
  -  I submitted an issue (#27862) earlier.
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
  -  Not necessary.
- [x] Did you write any new necessary tests?
  -  No.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

Hello @sanchit-gandhi, @patrickvonplaten, @anton-l, and @huggingface team, thanks for your efforts to bring Whisper to the awesome library, please help review this PR.

### CODE REVIEW

**Comment by nevikw39** [comment]
## Updated

I also found other audio classification models, such as wav2vec 2.0, HuBERT, WavLM, etc., except Audio Spectrogram Transformer, have similar issues. Therefore, I fixed them as well.

After applying the fix, `make fixup` failed for `data2vec_audio` despite the fact that `data2vec_text` and `data2vec_vision` support MSE for regression. `make fix-copies` seem to solve the problem by removing the fix of `Data2VecAudioForAudioFrameClassification`.

P.S. When rebasing the main branch from upstream, I made some mistakes and merge the main branch into my branch, which included others' commits into my branch. I reset the merge and push to my branch forcibly. Sorry for the mess.

**Comment by ArthurZucker** [comment]
Thanks, I'll ping @ylacombe and @sanchit-gandhi for a first review 😉

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for these fixes @nevikw39 and apologies for the late review! The code looks perfect - in the style of the Transformers loss functions 🙌 May I request that we implement a test for at least one of these models to confirm that we get the correct loss in the case that `config.num_labels>1`? Otherwise, this PR looks good to go!

File: src/transformers/models/data2vec/modeling_data2vec_audio.py [hunk 1]
```diff
@@ -1178,8 +1178,16 @@ def forward(
...
+            if self.config.num_labels > 1:
+                loss_fct = CrossEntropyLoss()
+                # move labels to correct device to enable PP
+                labels = labels.to(logits.device)
+                loss = loss_fct(logits.view(-1, self.config.num_labels), labels.view(-1))
+            else:
+                loss_fct = MSELoss()
+                # move labels to correct device to enable PP
+                labels = labels.to(logits.device)
+                loss = loss_fct(logits.view(-1), labels.view(-1))
```
Comment (ArthurZucker) [inline src/transformers/models/data2vec/modeling_data2vec_audio.py:1190]:
```suggestion
            # move labels to correct device to enable PP
            labels = labels.to(logits.device)
            if self.config.num_labels > 1:
                loss_fct = CrossEntropyLoss()
                loss = loss_fct(logits.view(-1, self.config.num_labels), labels.view(-1))
            else:
                loss_fct = MSELoss()
                loss = loss_fct(logits.view(-1), labels.view(-1))
```

**Review by ArthurZucker — APPROVED** [review]
LGTM, left a nit! 
Sorry for the late review

**Comment by nevikw39** [comment]
> Thanks for these fixes @nevikw39 and apologies for the late review! The code looks perfect - in the style of the Transformers loss functions 🙌 May I request that we implement a test for at least one of these models to confirm that we get the correct loss in the case that `config.num_labels>1`? Otherwise, this PR looks good to go!

Hi HuggingFace team,

Apologies for that I was busy in recent weeks.

I believe it's a great chance for me to learn to design Python unit testing. Could you please provide some kind instructions or examples for what the test should be? I checked existing testings and fount that only Whisper had tester for audio classification (with config `num_labels=2`), whereas Wav2Vec2, WavLM or HuBERT didn't.

Thank!

**Comment by ArthurZucker** [comment]
For testing I would recommend you to follow what is usually done in the code base so check that a loss is properly computed by simply passing dummy inputs and dummy labels! Make sure the CI is green as well an rebase on main as it hase been a while! 🤗

**Comment by nevikw39** [comment]
Hi HuggingFace team,

I resolved the suggestion of code review and the CI tests, and synced with the master branch (by merging rather than rebasing though I don't know whether it is appropriate, yet hopefully the PR would be squashed into a single commit, right?).

As for the test, I tried to take a close look on current unit tests. Sanchit (6th Jan.) said that we need "a test for at least one of these models to confirm that we get the correct loss in the case that `config.num_labels>1`". I fount that inside `tests/models/whisper/test_modeling_whisper.py`, class `WhisperEncoderModelTest` and `WhisperEncoderModelTester` seems to test the `forward()` method of `WhisperForAudioClassification` with [config `num_labels=2`](../tree/main/tests/models/whisper/test_modeling_whisper.py#L2597). So do these existing tests suffice?

For the remaining models, I haven't figured out how the existing tests deal with `num_labels` config. It appears that these models aren't covered by tests of forward, are they? Could we just leave those models as is?

Thanks!

**Comment by ArthurZucker** [comment]
> As for the test, I tried to take a close look on current unit tests. Sanchit (6th Jan.) said that we need "a test for at least one of these models to confirm that we get the correct loss in the case that config.num_labels>1". I fount that inside tests/models/whisper/test_modeling_whisper.py, class WhisperEncoderModelTest and WhisperEncoderModelTester seems to test the forward() method of WhisperForAudioClassification with [config num_labels=2](https://github.com/huggingface/transformers/tree/main/tests/models/whisper/test_modeling_whisper.py#L2597). So do these existing tests suffice?

yes it suffice if the `WhisperForAudioClassification` passes in that test! I can have a look and merge if it's alright!

**Review by ArthurZucker — COMMENTED** [review]
The test you are referring is this one? 
```python 
    def create_and_check_model_forward(self, config, inputs_dict, use_weighted_layer_sum=False):
        config.use_weighted_layer_sum = use_weighted_layer_sum
        model = WhisperForAudioClassification(config=config)
        model.to(torch_device).eval()

        input_features = inputs_dict["input_features"]

        with torch.no_grad():
            last_hidden_state = model(input_features).logits

        self.parent.assertTrue(last_hidden_state.shape, (13, 2))
```
if so, it needs to make sure the loss is MSELoss with an integration test 😉 not just the forward but loss if you give labels

**Comment by ArthurZucker** [comment]
Do you want me to add the test? 🤗

**Comment by nevikw39** [comment]
Hi Arthur,

Apologies for that I indefinitely postponed and just couldn’t find time to work on this task. I would really appreciate if you could help add the test. Thanks!

**Comment by ArthurZucker** [comment]
I don't have the bandwidth either, will let you handle it and the red ci! 🤗

**Pull request closed without merging**